pkg/client/cache/index.go

```diff
@@ -30,18 +30,43 @@ type Indexer interface {
 	Index(indexName string, obj interface{}) ([]interface{}, error)
 	// ListIndexFuncValues returns the list of generated values of an Index func
 	ListIndexFuncValues(indexName string) []string
+	// ByIndex lists object that match on the named indexing function with the exact key
+	ByIndex(indexName, indexKey string) ([]interface{}, error)
 }
 
 // IndexFunc knows how to provide an indexed value for an object.
-type IndexFunc func(obj interface{}) (string, error)
+type IndexFunc func(obj interface{}) ([]string, error)
+
+// IndexFuncToKeyFuncAdapter adapts an indexFunc to a keyFunc.  This is conversion can create errors when more than one
```

Comment:
I might add a comment that this is only useful if your IndexFunc indexes uniquely all incoming objects.  Many of the IndexFunc do not do this today, will have to check where you used this.